gcc/d/dfrontend/expression.c

```diff
@@ -4082,6 +4082,12 @@ Expression *ArrayLiteralExp::semantic(Scope *sc)
         return new ErrorExp();
     }
 
+    if (global.params.noTypeinfo && elements->dim != 0)
+    {
+        error (global.params.noTypeinfo, "Can't use array literal");
```

Comment:
> The problem here is that getTypeInfo might generate the TypeInfo for a type and this requires the TypeInfo declarations from object.d (it seems these are not needed for ctfe though). So we have to avoid calling getTypeInfo. We could also error in getTypeInfo but we can't provide good error messages in that case. I hoped I could return some dummy/error value and handle the error reporting in the visitor, but that doesn't work.

There aren't many `getTypeInfo` calls during semantic, and I'd guess they could/should all be moved to the glue layer.  Is there anything preventing that?  I think I had this on my DDMD list for a while but found an easier way to get things working.

> We could probably need a post-semantic phase ;-) I think it'd also be nice to have the druntime-call-generation shared between all compilers but that can't be done in semantic, AFAIK.

It can and it is in some cases (search for eg `_adSort`), but it shouldn't be.  Shouldn't be done during semantic that is, after is fine.